test/org/javarosa/core/TestBenchmark.java

```diff
@@ -0,0 +1,198 @@
+        private TreeElement dataRootNode;
+        private TreeElement savedRoot;
+        private FormDef formDef;
+
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+
+            FormParseInit formParseInit = new FormParseInit(r("nigeria_wards_external_combined.xml"));
+
+            FormEntryController formEntryController = formParseInit.getFormEntryController();
+
+            byte[] formInstanceAsBytes = null;
+            try {
+                formInstanceAsBytes = Files.readAllBytes(Paths.get(PathConst.getTestResourcePath().getAbsolutePath(), "populate-nodes-attributes-instance.xml"));
+            } catch (IOException e) {
+                fail("There was a problem with reading the test data.\n" + e.getMessage());
+            }
+            savedRoot = XFormParser.restoreDataModel(formInstanceAsBytes, null).getRoot();
+            formDef = formEntryController.getModel().getForm();
+            dataRootNode = formDef.getInstance().getRoot().deepCopy(true);
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class ParseWithExternalInstanceState {
+        Path resourcePath = PathConst.getTestResourcePath().toPath();
+
+        Path xFormFilePath = r("nigeria_wards_external.xml");
+        FormDef formDef;
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
+
+
+        }
+    }
+
+
+    @State(Scope.Thread)
+    public static class ParseWithInternalInstanceState {
+        Path resourcePath = PathConst.getTestResourcePath().toPath();
+
+        Path xFormFilePath = r("nigeria_wards_external_combined.xml");
+        FormDef formDef;
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
+
+
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class ExternalDataInstanceState {
+        ExternalDataInstance externalDataInstance = null;
+
+        @Setup(Level.Trial)
+        public void
+        initialize() {
+            Path resourcePath = PathConst.getTestResourcePath().toPath();
+            ReferenceManager.instance().addReferenceFactory((buildReferenceFactory("file", resourcePath.toString())));
+
+
+        }
+    }
+
+
+    @Benchmark
+    public void
+    benchmarkPopulateRootNode(BenchmarkState state, Blackhole bh) {
+        state.dataRootNode.populate(state.savedRoot, state.formDef);
+
+  }
+
+    @Benchmark
+    public void
+    benchmarkParseXFormWithExternalInstances(ParseWithExternalInstanceState state, Blackhole bh) {
+        try {
+            bh.consume(FormParserHelper.parse(state.xFormFilePath));
+        } catch (IOException e) {
```

Comment:
Could declaring an IOException in the method's signature be enough to make the test fail too? E.g.:

```java
@Benchmark
public void
benchmarkParseXFormWithInternalInstances(ParseWithInternalInstanceState state, Blackhole bh) throws IOException {
    bh.consume(FormParserHelper.parse(state.xFormFilePath));
}
```

Reply:
I think If the Exception is thrown, The Junit test would pass.

A Runtime Error is thrown in the catch block so as to make the test fail

Reply:
No problem. 

I can confirm that it would be enough. Tried this:

```java
@Benchmark
public void
benchmarkPopulateRootNode(BenchmarkState state, Blackhole bh) throws IOException {
    throw new IOException("some error");
}
```

Running this I get a junit test error and the full stacktrace in the output.

Reply:
Yes, Updated the code